programs/vest_api/src/vest_processor.rs

```diff
@@ -0,0 +1,394 @@
+//! vest program
+use crate::date_instruction::DateConfig;
+use crate::{
+    vest_instruction::{VestError, VestInstruction},
+    vest_state::VestState,
+};
+use bincode::deserialize;
+use chrono::prelude::*;
+use solana_config_api::get_config_data;
+use solana_sdk::{
+    account::{Account, KeyedAccount},
+    instruction::InstructionError,
+    pubkey::Pubkey,
+};
+
+fn parse_date_account(
+    keyed_account: &mut KeyedAccount,
+    expected_pubkey: &Pubkey,
+) -> Result<Date<Utc>, InstructionError> {
+    if keyed_account.account.owner != solana_config_api::id() {
+        return Err(InstructionError::IncorrectProgramId);
+    }
+
+    let account = parse_account(keyed_account, expected_pubkey)?;
+
+    let config_data =
+        get_config_data(&account.data).map_err(|_| InstructionError::InvalidAccountData)?;
+    let date_config =
+        deserialize::<DateConfig>(config_data).map_err(|_| InstructionError::InvalidAccountData)?;
+
+    Ok(date_config.dt.date())
+}
+
+fn parse_account<'a>(
+    keyed_account: &'a mut KeyedAccount,
+    expected_pubkey: &Pubkey,
+) -> Result<&'a mut Account, InstructionError> {
+    if keyed_account.unsigned_key() != expected_pubkey {
+        return Err(VestError::Unauthorized.into());
+    }
+
+    Ok(keyed_account.account)
+}
+
+fn parse_signed_account<'a>(
+    keyed_account: &'a mut KeyedAccount,
+    expected_pubkey: &Pubkey,
+) -> Result<&'a mut Account, InstructionError> {
+    if keyed_account.signer_key().is_none() {
+        return Err(InstructionError::MissingRequiredSignature);
+    }
+
+    parse_account(keyed_account, expected_pubkey)
+}
+
+pub fn process_instruction(
+    _program_id: &Pubkey,
+    keyed_accounts: &mut [KeyedAccount],
+    data: &[u8],
+) -> Result<(), InstructionError> {
+    let instruction = deserialize(data).map_err(|_| InstructionError::InvalidInstructionData)?;
+
+    match instruction {
+        VestInstruction::InitializeAccount {
+            terminator_pubkey,
+            payee_pubkey,
+            start_dt,
+            date_pubkey,
+            lamports,
+        } => {
+            let vest_state = VestState {
+                terminator_pubkey,
+                payee_pubkey,
+                start_dt,
+                date_pubkey,
+                lamports,
+                redeemed_lamports: 0,
+            };
+            vest_state.serialize(&mut keyed_accounts[0].account.data)
+        }
+        VestInstruction::RedeemTokens => {
+            let (ka0, ka1, ka2) = match keyed_accounts {
```

Comment:
How about descriptive variable names instead of indexes here so it's easier to keep track of which account is which?

Reply:
expanded, thanks!